llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -705,177 +625,253 @@ void EntryPointGroup::rebuildFromNames(const std::vector<std::string> &Names,
+        [](Function *F) -> std::string { return F->getName().str(); });
+    break;
+  case Scope_PerModule:
+    // The most complex case, because we should account for many other features
+    // like aspects used in a kernel, large-grf mode, reqd-work-group-size, etc.
+
+    // This is core of per-source device code split
+    Categorizer.registerSimpleStringAttributeRule(
+        sycl::utils::ATTR_SYCL_MODULE_ID);
+
+    // Optional features
+    Categorizer.registerSimpleFlagAttributeRule(
+        ::sycl::kernel_props::ATTR_LARGE_GRF, "large-grf");
+    Categorizer.registerListOfIntegersInMetadataSortedRule("sycl_used_aspects");
+    Categorizer.registerListOfIntegersInMetadataRule("reqd_work_group_size");
+    break;
+  default:
+    llvm_unreachable("Unexpected split scope");
+  }
+
+  // std::map is used here to ensure stable ordering of entry point groups,
+  // which is based on their contents, this greatly helps LIT tests
+  std::map<std::string, EntryPointSet> EntryPointsMap;
 
   // Only process module entry points:
-  for (auto &F : M.functions()) {
-    if (!isEntryPoint(F, EmitOnlyKernelsAsEntryPoints) ||
-        !MD.isEntryPointCandidate(F)) {
+  for (auto &F : MD.getModule().functions()) {
+    if (!isEntryPoint(F, EmitOnlyKernelsAsEntryPoints))
       continue;
-    }
 
-    auto Key = UsedOptionalFeatures(&F);
-    PropertiesToFunctionsMap[std::move(Key)].insert(&F);
+    std::string Key = Categorizer.computeCategoryFor(&F);
+    EntryPointsMap[std::move(Key)].insert(&F);
   }
 
-  if (PropertiesToFunctionsMap.empty()) {
+  EntryPointGroupVec Groups;
+
+  if (EntryPointsMap.empty()) {
     // No entry points met, record this.
     Groups.emplace_back(GLOBAL_SCOPE_NAME, EntryPointSet{});
   } else {
-    Groups.reserve(PropertiesToFunctionsMap.size());
-    for (auto &It : PropertiesToFunctionsMap) {
-      const UsedOptionalFeatures &Features = It.first;
-      EntryPointSet &EntryPoints = It.second;
-
-      // Start with properties of a source module
-      EntryPointGroup::Properties MDProps = MD.getEntryPointGroup().Props;
-      // Propagate LargeGRF flag to entry points group
-      if (Features.UsesLargeGRF)
-        MDProps.UsesLargeGRF = true;
-      Groups.emplace_back(
-          Features.generateModuleName(MD.getEntryPointGroup().GroupId),
-          std::move(EntryPoints), MDProps);
-    }
+    Groups.reserve(EntryPointsMap.size());
+    // Start with properties of a source module
+    EntryPointGroup::Properties MDProps = MD.getEntryPointGroup().Props;
+    for (auto &It : EntryPointsMap)
+      Groups.emplace_back(It.first, std::move(It.second), MDProps);
   }
 
-  if (Groups.size() > 1)
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
+
+  if (DoSplit)
     return std::make_unique<ModuleSplitter>(std::move(MD), std::move(Groups));
   else
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
```

Comment:
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(MD), std::move(Groups));

  return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
```

Reply:
Thanks, applied in aaf1e92b82415656faec197d9afa340fe54627bd